edb/edgeql/compiler/expr.py

```diff
@@ -646,7 +647,34 @@ def compile_GlobalExpr(
 def compile_TypeCast(
     expr: qlast.TypeCast, *, ctx: context.ContextLevel
 ) -> irast.Set:
-    target_stype = typegen.ql_typeexpr_to_type(expr.type, ctx=ctx)
+    try:
+        target_stype = typegen.ql_typeexpr_to_type(expr.type, ctx=ctx)
+    except errors.InvalidReferenceError as e:
+        if (
+            e.hint is None
+            and isinstance(expr.type, qlast.TypeName)
+            and isinstance(expr.type.maintype, qlast.ObjectRef)
+        ):
+            type_name: str = expr.type.maintype.name
+            type_module: Optional[str] = expr.type.maintype.module
+
+            name: sn.Name
+            if type_module is not None:
+                name = sn.QualName(type_module, type_name)
+            else:
+                name = sn.UnqualName(type_name)
```

Comment:
There's a function in schema.utils (I think?) for converting from an ObjectRef to a Name